WordPress/Classes/Networking/LoginClient.swift

```diff
@@ -82,94 +89,91 @@ actor LoginClient {
         ])
     }
 
-    @MainActor
-    static func displaySelfHostedLoginView(in navigationController: UINavigationController) {
-        let loginClient = LoginClient(session: .shared)
-        let vm = SelfHostedLoginViewModel(loginClient: loginClient)
-        let loginViewController = UIHostingController(rootView: LoginWithUrlView(viewModel: vm))
-
-        vm.onLoginComplete = { [weak loginViewController] in
-            loginViewController?.dismiss(animated: true)
+    func login(site: String, from anchor: ASPresentationAnchor?) async -> Result<WordPressOrgCredentials, LoginClientError> {
+        let authURL = await buildLoginUrl(from: site)
+        switch authURL {
+        case let .success(authURL):
+            let urlWithToken = await presentWebView(authURL: authURL, from: anchor)
+            switch urlWithToken {
+            case let .success(urlWithToken):
+                return await handleAuthenticationCallback(urlWithToken)
+            case let .failure(error):
+                return .failure(error)
+            }
+        case let .failure(error):
+            return .failure(error)
         }
-
-        navigationController.pushViewController(loginViewController, animated: true)
     }
-}
-
-class SelfHostedLoginViewModel: LoginWithUrlView.ViewModel {
-    private let loginClient: LoginClient
-    private let presentationContextProvider = SelfHostedLoginViewModelAuthenticationContextProvider()
-
-    var onLoginComplete: (() -> Void)?
 
-    init(loginClient: LoginClient) {
-        self.loginClient = loginClient
-        super.init()
+    @MainActor
+    func buildLoginUrl(from userInput: String) async -> Result<URL, LoginClientError> {
```

Comment:
This whole method should be pushed down into the shared layer – other platforms will need it too